File: contest/LeetCode_Weekly_Contest_204/Number_of_Ways_to_Reorder_Array_to_Get_Same_BST.py
class Solution(object):
    MOD = 1000000007
    
    def dfs(self, nums):
        if len(nums) <= 1: return 1
        
        a, b = [], []
        for x in nums:
            if x == nums[0]: 
                continue
            if x < nums[0]: 
                a.append(x)
            else:
                b.append(x)
                
        return (self.f[len(a)][len(b)] * self.dfs(a) * self.dfs(b)) % self.MOD
        
    def numOfWays(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        
        self.n = len(nums)
        
        self.f = [[0 for _ in xrange(self.n + 1)] for _ in xrange(self.n + 1)]
        for i in xrange(self.n+1):
            self.f[i][0] = self.f[0][i] = 1
            
        for i in xrange(1, self.n+1):
            for j in xrange(1, self.n+1):
                self.f[i][j] = (self.f[i-1][j] + self.f[i][j-1]) % self.MOD
                
        ret = self.dfs(nums)
        return (ret - 1 + self.MOD) % self.MOD

File: contest/LeetCode_Weekly_Contest_204/Number_of_Ways_to_Reorder_Array_to_Get_Same_BST.cpp
const int MOD = 1e9+7;
const int N = 1005;

int f[N][N];

class Solution {
public:
    int dfs(const vector<int>& nums) {
        if (nums.size() <= 1) return 1;
        vector<int> a, b;
        int root = nums.front();
        int n = nums.size();
        for (int i = 1; i < n; ++i) {
            if (nums[i] < root) a.push_back(nums[i]);
            else b.push_back(nums[i]);
        }
        
        int acnt = a.size();
        int bcnt = b.size();
        
        long long sum = f[acnt][bcnt];
        sum = (sum * dfs(a)) % MOD;
        sum = (sum * dfs(b)) % MOD;
        
        return sum;
    }
    
    int numOfWays(vector<int>& nums) {
        int n = nums.size();
        
        memset(f, 0, sizeof f);
        
        for (int i = 1; i <= n; ++i)
            f[i][0] = f[0][i] = 1;
        
        for (int i = 1; i <= n; ++i)
            for (int j = 1; j <= n; ++j) {
                f[i][j] = (f[i][j-1] + f[i-1][j]) % MOD;
            }
        
        int ret = dfs(nums);
        return (ret - 1 + MOD) % MOD;
    }
};
